[linter] Updating RuleID and generalizing related description missing rules

vishrutshah opened the pull request:
Fixes #1919 
Fixes https://github.com/Azure/autorest/issues/2044

veronicagg commented on src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs:
```diff
@@ -8,30 +8,23 @@
                 if (schema.Description == null)
                 {
-                    yield return new ValidationMessage(new FileObjectPath(context.File, context.Path), this, "description");
+                    yield return new ValidationMessage(new FileObjectPath(context.File, context.Path), this, "model type");
```
could we get the name of the model type in the message?

veronicagg commented on src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs:
```diff
@@ -8,30 +8,23 @@
-            if (schema != null && string.IsNullOrEmpty(schema.Reference) && schema.RepresentsCompositeType())
+            if (string.IsNullOrWhiteSpace(schema?.Reference) && schema.RepresentsCompositeType())
             {
                 if (schema.Description == null)
```
should this condition also consider if description is empty or whitespace, similar to the other ones?

vishrutshah replied:
fixed

veronicagg commented on src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs:
```diff
@@ -1,34 +1,24 @@
+        public override bool IsValid(Operation entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format("operation '{0}'", entity.OperationId) };
+            return entity?.Description != null;
```
should this condition be the same as the one for "parameters" "return entity?.Description != null || !string.IsNullOrWhiteSpace(entity.Reference);"

vishrutshah replied:
I am not sure what is `entity.Reference` in the context of the operation. I should definitely check for whitespace though

vishrutshah replied:
Updated it to ` return !string.IsNullOrWhiteSpace(entity?.Description);`

veronicagg commented on src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs:
```diff
@@ -1,34 +1,24 @@
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(SwaggerParameter entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { "parameter" };
```
could we get the parameter name in the message?

dsgouda replied:
I'm guessing you're using this validation rule multiple times in the spec. It's usually more efficient to set this as a readonly static variable instead so we don't create the same object multiple times unnecessarily

vishrutshah replied:
sure

dsgouda commented on src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs:
```diff
@@ -8,30 +8,23 @@
         public override IEnumerable<ValidationMessage> GetValidationMessages(Schema schema, RuleContext context)
         {
-            if (schema != null && string.IsNullOrEmpty(schema.Reference) && schema.RepresentsCompositeType())
+            if (string.IsNullOrWhiteSpace(schema?.Reference) && schema.RepresentsCompositeType())
```
You will still want to do ```schema?.RepresentsCompositeType() ==true```

dsgouda commented on src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs:
```diff
@@ -8,30 +8,23 @@
+            if (string.IsNullOrWhiteSpace(schema?.Reference) && schema.RepresentsCompositeType())
             {
-                if (schema.Description == null)
+                if (string.IsNullOrWhiteSpace(schema.Description))
```
```schema?.Description```

vishrutshah replied:
After having 

`if (string.IsNullOrWhiteSpace(schema?.Reference) && schema?.RepresentsCompositeType() == true)` will i get null schema inside?

dsgouda replied:
Ahh, fair point

vishrutshah commented:
@veronicagg I looked at the PR and saw that we were not applying the rule for the properties of the model. Now it's corrected by applying `ModelTypeIncoplete` rule to model properties as well. Also added test in the same swagger. 

Here is the output for that PR with removed one description inside mode.

```
  {
    "type": "Warning",
    "code": "ModelTypeIncomplete",
    "message": "This 'availableClusterVersions' model/property lacks 'description' property. Consider adding a 'description' element. Accurate description is essential for maintaining reference documentation.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterProperties/properties",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterProperties/properties",
    "id": "M4000",
    "validationCategory": "SDKViolation",
    "providerNamespace": null,
    "resourceType": null
  },
  {
    "type": "Warning",
    "code": "ModelTypeIncomplete",
    "message": "This 'value' model/property lacks 'description' property. Consider adding a 'description' element. Accurate description is essential for maintaining reference documentation.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterListResult/properties",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterListResult/properties",
    "id": "M4000",
    "validationCategory": "SDKViolation",
    "providerNamespace": null,
    "resourceType": null
  }
```

Also addressed the comments. @veronicagg / @dsgouda please look at the PR again. Thanks

dsgouda commented on src/modeler/AutoRest.Swagger/Validation/DescriptionRequired.cs:
```diff
@@ -0,0 +1,36 @@
+        /// <remarks>
+        /// This may contain placeholders '{0}' for parameterized messages.
+        /// </remarks>
+        public override string MessageTemplate => Resources.MissingDescription;
```
If this is going to get overridden in the base classes anyway, no need to declare here

vishrutshah replied:
as discussed offline, this is the base class

dsgouda commented on src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs:
```diff
@@ -8,30 +8,30 @@
-                if (schema.Description == null)
+                string key = definition.Key;
+                Schema schema = definition.Value;
+                if (string.IsNullOrWhiteSpace(schema?.Reference))
```
Why do we need to check for ```schema.Reference``` here?

dsgouda commented on src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs:
```diff
@@ -1,34 +1,26 @@
+        public override bool IsValid(SwaggerParameter entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format(ParameterTypeFormatter, entity.Name) };
+            return !string.IsNullOrWhiteSpace(entity?.Description) || !string.IsNullOrWhiteSpace(entity.Reference);
```
nit:The rule won't be called for a null entity, you can say ```entity.Description```

vishrutshah replied:
`Reference` cannot be in `SwaggerParameter` schema from `http://swagger.io/specification/#parameterObject`

vishrutshah replied:
@veronicagg I found that `parameters` in the operations can have refs. So i cannot remove this. 

ex: https://github.com/Azure/autorest/blob/master/src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/positive/clean-complex-spec.json#L166

vishrutshah commented on src/core/AutoRest.Core/Properties/Resources.resx:
```diff
@@ -283,7 +283,7 @@ Modelers:
   </data>
   <data name="MissingDescription" xml:space="preserve">
-    <value>Consider adding a 'description' element, essential for maintaining reference documentation.</value>
+    <value>This {0} lacks 'description' property. Consider adding a 'description' element. Accurate description is essential for maintaining reference documentation.</value>
```
remove `this`

vishrutshah commented on src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs:
```diff
@@ -1,34 +1,24 @@
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(Operation entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format("operation '{0}'", entity.OperationId) };
```
caps on `O`

vishrutshah commented:
@azuresdkci Test this please

vishrutshah commented:
@azuresdkci Test this please

dsgouda commented on src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs:
```diff
@@ -1,34 +1,24 @@
+        public override bool IsValid(Operation entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format("Operation '{0}'", entity.OperationId) };
+            return !string.IsNullOrWhiteSpace(entity?.Description);
```
Guess you missed ```entity?.Description```

vishrutshah commented:
Thanks for the review!!
